### PR TITLE
Augmentation pipeline

### PR DESCRIPTION
**Pull request by dodecaphony**

### CODE REVIEW

File: pipeline_demo.py
Comment (meduzick) [inline pipeline_demo.py:1]: Pls, make a similar notebook and put into sage/notebooks

File: sage/pipeline/augmentors.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+from abc import ABC, abstractmethod
+from sage.spelling_corruption.configuration_corruptor import CharAugConfig, WordAugConfig, SBSCConfig
+from sage.spelling_corruption.corruptor import CharAugCorruptor, WordAugCorruptor, SBSCCorruptor
+from typing import Optional
```
Comment (meduzick) [inline sage/pipeline/augmentors.py]: Not sure, but import from conventional typing may be better placed before import from sage package

File: sage/pipeline/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+from typing import Optional, List
+import random
```
Comment (meduzick) [inline sage/pipeline/pipeline.py:1]: import random has to be before import from typing I guess

**Review by meduzick — CHANGES_REQUESTED** [review]
Just a few more details, everything else is ok

File: sage/pipeline/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+from .config import PipelineConfig
+from .augmentors import CharAugmentor, WordAugmentor, SBSCorruptor
+from sage.spelling_corruption.configuration_corruptor import CharAugConfig, WordAugConfig, SBSCConfig
+
+
+class AugmentationPipeline:
+    def __init__(self, config: PipelineConfig = PipelineConfig(), shuffle: bool = True):
+        self.augmentors = []
+        self.config = config
+        self._add_all_augmentors()
```
Comment (danil31219as) [inline sage/pipeline/pipeline.py]: You always add all the augmentors

File: tests/test_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            word_unit_prob=0.3,
+            sbsc_lang="ru",
+            sbsc_reference_dataset_name_or_path=DatasetsAvailable.MedSpellchecker.name,
+            sbsc_reference_dataset_split="test"
+        )
+        self.sample_text = "Заметьте, не я это предложил!"
+
+    def test_char_augmentation(self):
+        pipeline = AugmentationPipeline(config=self.pipeline_config, shuffle=False)
+        pipeline.add_char_augmentor()
```
Comment (danil31219as) [inline tests/test_pipeline.py]: You've already added all kinds of augmentors

File: tests/test_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        pipeline.add_word_augmentor()
+        pipeline.add_sbsc_augmentor()
+        pipeline.set_order([0, 1, 2])  # Все аугментаторы
+        augmented_text = pipeline.augment(self.sample_text, seed=1)
+        self.assertNotEqual(self.sample_text, augmented_text)
+        print(f"All Augmentations Result: {augmented_text}")
+
+    if __name__ == "__main__":
+        unittest.main()
```
Comment (danil31219as) [inline tests/test_pipeline.py:51]: Extra indentation, tests don't run, how did you test this code?

File: sage/pipeline/config.py [hunk 6]
```diff
@@ -28,3 +29,44 @@ def set_sbsc_params(self, lang: str, dataset_name_or_path: str, dataset_split: s
...
         self.sbsc_reference_dataset_split = dataset_split
+
+    def __set_language(self, lang: str):
+        """
+        Sets the language and corresponding dataset path based on the provided language.
+
+        Args:
+            lang (str): The language code ('ru' or 'en').
+        """
+        self.sbsc_lang = lang
```
Comment (danil31219as) [inline sage/pipeline/config.py:40]: duplicates __init__ method

File: sage/pipeline/config.py [hunk 7]
```diff
@@ -1,18 +1,19 @@
...
+    def __init__(self, lang: str = 'ru'):
+        self.char_min_aug: int = 1
+        self.char_max_aug: int = 3
+        self.char_unit_prob: float = 0.2
+        self.word_min_aug: int = 1
+        self.word_max_aug: int = 3
+        self.word_unit_prob: float = 0.3
+        self.sbsc_lang: str = lang
+        self.sbsc_reference_dataset_name_or_path: str = DatasetsAvailable.MedSpellchecker.name
+        self.sbsc_reference_dataset_split: str = "test"
```
Comment (danil31219as) [inline sage/pipeline/config.py:15]: Why is lang an argument, but everything else is fixed?

File: sage/pipeline/config.py [hunk 8]
```diff
@@ -28,3 +29,44 @@ def set_sbsc_params(self, lang: str, dataset_name_or_path: str, dataset_split: s
...
+
+        Args:
+            lang (str): The language code ('ru' or 'en').
+        """
+        self.sbsc_lang = lang
+        if lang == 'en':
+            base_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', '..'))
+            self.sbsc_reference_dataset_name_or_path = os.path.join(base_dir, "data", "example_data", "jfleg")
+        elif lang != 'ru':
+            raise ValueError("Unsupported language. Supported languages are 'ru' and 'en'.")
```
Comment (danil31219as) [inline sage/pipeline/config.py:45]: not scalable, but not critical

File: sage/pipeline/pipeline.py [hunk 9]
```diff
@@ -1,57 +1,101 @@
...
-    def remove_augmentor(self, augmentor_type):
-        self.augmentors = [augmentor for augmentor in self.augmentors if not isinstance(augmentor, augmentor_type)]
+    def remove_augmenter(self, augmenter_type):
+        """
+        Removes all instances of the specified augmenter type from the pipeline.
+
+        Args:
+            augmenter_type (type): The type of augmenter to remove.
+        """
+        self.augmenters = [augmenter for augmenter in self.augmenters if not isinstance(augmenter, augmenter_type)]
```
Comment (danil31219as) [inline sage/pipeline/pipeline.py:77]: When I work with the AugmentationPipeline class and its methods like add_char_augmenter, I don't know about the existence of augmenter types, but I have to get them somewhere

File: sage/pipeline/pipeline.py [hunk 10]
```diff
@@ -1,57 +1,101 @@
...
+        Initializes the AugmentationPipeline with a given configuration and optional shuffling.
+
+        Args:
+            config (PipelineConfig): The configuration object containing settings for the augmenters.
+            shuffle (bool): Whether to shuffle the order of augmenters. Default is True.
+        """
+        self.augmenters = []
         self.config = config
-        self._add_all_augmentors()
+        self._add_all_augmenters()
```
Comment (danil31219as) [inline sage/pipeline/pipeline.py:19]: why?

File: sage/pipeline/pipeline.py [hunk 11]
```diff
@@ -1,57 +1,101 @@
...
+        self.add_word_augmenter()
+        self.add_sbsc_augmenter()
 
-    def _shuffle_augmentors(self):
-        random.shuffle(self.augmentors)
+    def _shuffle_augmenters(self):
+        """
+        Randomly shuffles the order of the augmenters in the pipeline.
+        """
+        random.shuffle(self.augmenters)
```
Comment (danil31219as) [inline sage/pipeline/pipeline.py:35]: seed?

File: sage/pipeline/pipeline.py [hunk 12]
```diff
@@ -1,57 +1,101 @@
...
             lang=self.config.sbsc_lang,
             reference_dataset_name_or_path=self.config.sbsc_reference_dataset_name_or_path,
             reference_dataset_split=self.config.sbsc_reference_dataset_split
         )
-        self.augmentors.append(SBSCorruptor(sbsc_config))
+        self.augmenters.append(SBSCorruptor(sbsc_config))
 
-    def remove_augmentor(self, augmentor_type):
-        self.augmentors = [augmentor for augmentor in self.augmentors if not isinstance(augmentor, augmentor_type)]
+    def remove_augmenter(self, augmenter_type):
```
Comment (danil31219as) [inline sage/pipeline/pipeline.py:70]: type of augmenter_type?

File: tests/test_pipeline.py [hunk 13]
```diff
@@ -4,52 +4,46 @@
...
-        pipeline.add_char_augmentor()
-        pipeline.add_word_augmentor()
-        pipeline.add_sbsc_augmentor()
-        pipeline.set_order([0, 1, 2])  # Все аугментаторы
-        augmented_text = pipeline.augment(self.sample_text, seed=1)
+        pipeline = AugmentationPipeline(config=self.pipeline_config, shuffle=True)
+        pipeline.add_char_augmenter()
+        pipeline.add_word_augmenter()
+        pipeline.add_sbsc_augmenter()
+        pipeline.set_order([0, 1, 2])  # All augmenters
```
Comment (danil31219as) [inline tests/test_pipeline.py:45]: maybe shuffled order?

**Review by danil31219as — CHANGES_REQUESTED** [review]

**Pull request merged**